File: clients/go/vectorembed/http.go
package vectorembed

import (
	"encoding/json"
	"math/rand"
	"strconv"
	"sync"
	"time"
)

// Retry configuration constants.
const (
	DefaultRetries   = 3
	DefaultBaseDelay = 1 * time.Second
	DefaultMaxDelay  = 30 * time.Second
	JitterFactor     = 0.2 // ±20%
)

// retryableStatuses maps HTTP status codes that should trigger a retry.
var retryableStatuses = map[int]bool{
	429: true, // Too Many Requests
	500: true, // Internal Server Error
	502: true, // Bad Gateway
	503: true, // Service Unavailable
	504: true, // Gateway Timeout
}

// lockedRand provides thread-safe random number generation.
type lockedRand struct {
	mu   sync.Mutex
	rand *rand.Rand
}

// Float64 returns a random float64 in [0, 1) in a thread-safe manner.
func (r *lockedRand) Float64() float64 {
	r.mu.Lock()
	defer r.mu.Unlock()
	return r.rand.Float64()
}

// globalRand is the package-level thread-safe random generator.
var globalRand = &lockedRand{
	rand: rand.New(rand.NewSource(time.Now().UnixNano())),
}

// calculateDelay returns the delay for a retry attempt using exponential backoff with jitter.
// The base delay is doubled for each attempt, then jitter of ±JitterFactor is applied.
// The result is capped at DefaultMaxDelay (plus jitter).
func calculateDelay(attempt int) time.Duration {
	// Calculate base delay with exponential backoff: base * 2^attempt
	delay := DefaultBaseDelay * time.Duration(1<<uint(attempt))

	// Cap at max delay
	if delay > DefaultMaxDelay {
		delay = DefaultMaxDelay
	}

	// Apply jitter: delay * (1 + random(-JitterFactor, +JitterFactor))
	jitter := (globalRand.Float64()*2 - 1) * JitterFactor // Random in [-JitterFactor, +JitterFactor]
	delay = time.Duration(float64(delay) * (1 + jitter))

	return delay
}

// isRetryable returns true if the given HTTP status code should trigger a retry.
func isRetryable(statusCode int) bool {
	return retryableStatuses[statusCode]
}

// errorResponse represents the JSON error response from the API.
type errorResponse struct {
	Error string `json:"error"`
}

// mapStatusToError maps an HTTP status code to the appropriate error type.
// The body parameter contains the response body which may contain an error message.
// The headers parameter may contain a Retry-After header for rate limit errors.
func mapStatusToError(statusCode int, body []byte, headers map[string][]string) error {
	// Try to extract error message from JSON body
	var errResp errorResponse
	message := ""
	if err := json.Unmarshal(body, &errResp); err == nil && errResp.Error != "" {
		message = errResp.Error
	}

	// Set default messages if not extracted from body
	if message == "" {
		switch {
		case statusCode == 400:
			message = "bad request"
		case statusCode == 401:
			message = "authentication failed"
		case statusCode == 429:
			message = "rate limit exceeded"
		case statusCode >= 500:
			message = "server error"
		default:
			message = "unknown error"
		}
	}

	baseErr := APIError{
		Message:    message,
		StatusCode: statusCode,
	}

	switch {
	case statusCode == 400:
		return &ValidationError{APIError: baseErr}
	case statusCode == 401:
		return &AuthenticationError{APIError: baseErr}
	case statusCode == 429:
		retryAfter := time.Duration(0)
		if headers != nil {
			if vals, ok := headers["Retry-After"]; ok && len(vals) > 0 {
				retryAfter = parseRetryAfter(vals[0])
			}
		}
		return &RateLimitError{
			APIError:   baseErr,
			RetryAfter: retryAfter,
		}
	case statusCode >= 500:
		return &ServerError{APIError: baseErr}
	default:
		return &APIError{
			Message:    message,
			StatusCode: statusCode,
		}
	}
}

// parseRetryAfter parses a Retry-After header value and returns the duration.
// Returns 0 if the header is empty or cannot be parsed.
// Note: This implementation only supports seconds format, not HTTP-date format.
func parseRetryAfter(header string) time.Duration {
	if header == "" {
		return 0
	}
	seconds, err := strconv.Atoi(header)
	if err != nil {
		return 0
	}
	return time.Duration(seconds) * time.Second
}

File: clients/python/src/vector_embed_client/http.py
"""HTTP layer with retry logic for vector-embed-cache-client."""

from __future__ import annotations

import contextlib
import random
import sys
from typing import Any

from vector_embed_client.errors import (
    AuthenticationError,
    RateLimitError,
    ServerError,
    ValidationError,
    VectorEmbedError,
)

VERSION = "0.1.0"

RETRY_CONFIG: dict[str, Any] = {
    "base_delay": 1.0,  # 1 second
    "max_delay": 30.0,  # 30 seconds cap
    "jitter_factor": 0.2,  # ±20% randomization
    "retryable_statuses": [429, 500, 502, 503, 504],
}


def calculate_delay(attempt: int) -> float:
    """Calculate delay for retry attempt with exponential backoff and jitter.

    Args:
        attempt: The retry attempt number (0-indexed).

    Returns:
        Delay in seconds with jitter applied.
    """
    base_delay: float = RETRY_CONFIG["base_delay"]
    max_delay: float = RETRY_CONFIG["max_delay"]
    jitter_factor: float = RETRY_CONFIG["jitter_factor"]

    # Exponential backoff: base_delay * 2^attempt
    delay: float = base_delay * (2**attempt)

    # Cap at max_delay
    delay = min(delay, max_delay)

    # Apply jitter: ±jitter_factor
    jitter: float = delay * jitter_factor * (2 * random.random() - 1)
    delay = delay + jitter

    return delay


def map_status_to_error(
    status: int,
    body: dict[str, Any],
    headers: dict[str, str],
) -> VectorEmbedError:
    """Map HTTP status code to appropriate exception.

    Args:
        status: HTTP status code.
        body: Response body as dict.
        headers: Response headers.

    Returns:
        Appropriate VectorEmbedError subclass instance.
    """
    message = body.get("error", "Unknown error")

    if status == 400:
        return ValidationError(
            message=message,
            code="validation_error",
            status=status,
        )
    elif status == 401:
        return AuthenticationError(
            message=message,
            code="authentication_error",
            status=status,
        )
    elif status == 429:
        retry_after: int | None = None
        retry_after_header = headers.get("retry-after")
        if retry_after_header is not None:
            with contextlib.suppress(ValueError):
                retry_after = int(retry_after_header)
        return RateLimitError(
            message=message,
            code="rate_limit_error",
            status=status,
            retry_after=retry_after,
        )
    elif status >= 500:
        return ServerError(
            message=message,
            code="server_error",
            status=status,
        )
    else:
        return VectorEmbedError(
            message=message,
            code="unknown_error",
            status=status,
        )


def build_headers(api_key: str) -> dict[str, str]:
    """Build HTTP request headers.

    Args:
        api_key: API key for authentication.

    Returns:
        Dict of headers to include in requests.
    """
    python_version = f"{sys.version_info.major}.{sys.version_info.minor}"
    user_agent = f"vector-embed-cache-client/{VERSION} python/{python_version}"

    return {
        "Authorization": f"Bearer {api_key}",
        "Content-Type": "application/json",
        "Accept": "application/json",
        "User-Agent": user_agent,
    }
